### DISCUSSION TITLE
Can't connect a collection to Slack, oAuth stuck in a redirect loop.

### DISCUSSION THREAD

**Discussion by pankaj9296** [outline/outline#4993] [Self Hosting] [answered]
When I try to connect a collection to Slack under Integrations > Slack, it redirects you to Slack for oAuth and redirects back to the wiki, and stuck on a redirect loop on the following URL:
https://wiki.my-domain.com/auth/slack.post?code=a-long-string&state=another-string

The same happens when you click the 'Add To Slack' command under Integrations > Slack.
It is stuck in a redirect loop on the following URL:
https://wiki.my-domain.com/auth/slack.commands?code=long-code&state=another-long-code

I don't see much on the console logs except the following printed for each redirect in the loop:
```
{"label":"http","level":"info","message":"  --> GET /auth/slack.commands?code=some-code&state=some-code 200 8ms 253b"}
{"label":"http","level":"info","message":"  <-- GET /auth/slack.commands?code=some-code&state=some-code"}
```

FYI, the slack login works well.
I have the following env variables set already:
SLACK_VERIFICATION_TOKEN='xyz'
SLACK_APP_ID='xyz'
SLACK_MESSAGE_ACTIONS=true
SLACK_CLIENT_ID='xyz'
SLACK_CLIENT_SECRET='xyz'

The actual code/tokens has been replaced with the placeholders above.

**To Reproduce**
1. Login to Outline and Go to 'Integrations > Slack'
2. Click on 'Add To Slack'
3. Click allow on Oauth Window of slack
4. It redirects to self hosted outline and stuck in redirect loop.

- Install: self-hosted
- Version: bef96735300c55479add2745b60667f505eb0276

 - OS: Mac OS 13.2.1
 - Browser Chrome
 - Version 110.0.5481.177 (Official Build) (arm64)

**Comment by tommoor** [comment]
What is the redirect loop?

**Comment by pankaj9296** [comment]
@tommoor 
So when I click the Add To Slack button, it redirects to Slack, when I allow permissions from Slack, it redirects back to outline URL https://wiki.domain.com/auth/slack.commands?... and keeps reloading the page on the same URL. Please check the attached video.
https://user-images.githubusercontent.com/17493609/223198162-15cabe3e-64fc-43ee-9d93-8340a222bb3f.mp4

**Reply by tommoor** [reply]
The server thinks you're logged out as it can't load the session cookie for some reason, that's the best clue I can offer based on that video.

**Comment by macgate** [comment]
I have the same problem. I didn't find the solution for now.
See this post https://github.com/outline/outline/discussions/3169
I think this is because at the installation of the outline, i configured it for Google and not for Slack, then after all function for Slack are broken.

**Comment by marc0777** [comment]
I'm too encountering the same issue. Slack looks configured correctly: login, Slack command and already configured notifications are all working. I'm running v68.1, and I'm sure that this was working at least up until v66.3.

I can provide further information/testing if needed.

**Comment by attakei** [comment, marked as answer]
I have same problem and resolved with small patch in my environment.
Please see [commit on forked](https://github.com/outline/outline/compare/main...attakei:outline:fix/refer-state-in-slack) that is based from patched js code.

At current slack plugin, endpoint of redirect_url from slack refer ctx.state.user to check authenticated.
But authenticatation process save into ctx.state.auth, it seems that ctx.state.user is always undefined and response renders [`redirectOnClient()`](https://github.com/outline/outline/blob/main/plugins/slack/server/auth/slack.ts#L54).

If [this commit](https://github.com/outline/outline/compare/main...attakei:outline:fix/refer-state-in-slack) is correctlly as patch for resolve issue, I can create pull-request asap.

**Reply by tommoor** [reply]
You're totally write, I pushed the fix. Somehow the cloud-hosted fork had deviated on these lines hence why this wasn't an issue on `getoutline.com`.
